include/aws/common/trace_event.h

```diff
@@ -0,0 +1,84 @@
+#ifndef AWS_COMMON_TRACE_EVENT_H
+#define AWS_COMMON_TRACE_EVENT_H
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+/* Disable warnings with windows build */
+#ifdef _MSC_VER
+#    pragma warning(disable : C4996)
+#endif
+
+#include <aws/common/common.h>
+
+AWS_EXTERN_C_BEGIN
+
+#define AWS_TRACE_EVENT_BEGIN(category, name) aws_trace_event_new(category, name, EVENT_PHASE_BEGIN)
+
+#define AWS_TRACE_EVENT_END(category, name) aws_trace_event_new(category, name, EVENT_PHASE_END)
+
+#define AWS_TRACE_EVENT_INSTANT(category, name) aws_trace_event_new(category, name, EVENT_PHASE_INSTANT)
+
+// Phase macros
+//! add more phase types later as the app progresses
+#define EVENT_PHASE_BEGIN ('B')
+#define EVENT_PHASE_END ('E')
+#define EVENT_PHASE_INSTANT ('I')
+
+#define AWS_TRACE_EVENT_TIME_DISPLAY_MICRO 0
+#define AWS_TRACE_EVENT_TIME_DISPLAY_NANO 1
+
+struct aws_trace_event_metadata {
+    /* should be B/E for same scope or S/F for outside of scope */
+    char phase;
+    /* name of the event */
+    char name[15];
+    /* category of the event */
+    char category[15];
+    /* timestamp in milliseconds */
+    uint64_t timestamp;
+    uint64_t thread_id;
+    int process_id;
+    /* args for more metadata to be added later */
+};
+
+/*
+ * MUST CALL TO START TRACE SYSTEM
+ * Starts the aws_message_bus in a background thread and subscribes the listener to it.
+ * Initializes a JSON object to store trace event data
+ * Must be called before using event_trace_add or close_event_trace
+ *
+ */
+AWS_COMMON_API
+int aws_trace_system_init(struct aws_allocator *allocator);
+
+/*
+ * MUST CALL TO AVOID MEMORY LEAKS
+ * Cleans up the aws_message_bus and unsubscribes the listener.
+ * Prints the JSON object that stores trace event data to the given filename
+ * as a .JSON file.
+ * Closes the JSON object and frees all allocated memory.
+ * Code 0 to not write out file and Code 1 to write out to file
+ * time_unit how seconds are displayed on chrome://tracing
+ */
+AWS_COMMON_API
+void aws_trace_system_clean_up(int code, int time_unit, const char *filename);
```

Comment:
tldr; `aws_trace_system_init(struct aws_allocator *allocator, const char *filename)`

These arguments should go into the init() function, not the clean_up(). Currently, you're writing everything out during clean_up(), but eventually we want to be writing to the file incrementally as events roll in (so we don't have an infinitely growing JSON object in memory) so we'll need to know the filename up front.

also, maybe let's get rid of the `code` argument. "int code" isn't very descriptive compared to something like`bool write_to_file`. But also, we could just say that passing NULL to `filename` is how the user says they don't want to a file, so we use one argument that does what it says instead of 2 different arguments that influence the same thing.

also, let's drop the `time_unit` option for now, just pick one. I imagine there will be LOTS of configuration options for how we want this thing to run when all is said and done. It's a pain to add arguments to a function over time, it breaks the API and you need to track down everywhere that function is called and update how it's called...
Let's just keep the init() call simple for now, and hard-code misc options like time-unit. Later, when we're more certain what to expose, we can introduce some alternate init() function that takes a million arguments, or takes a struct where the million arguments are members of the struct.